File: 2ndSemester/AlgorithimsAndDataStructuresII/Practice/PraticalAssignment01/Q07-IsC/Is.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

#define MAXTAM 1000

bool checkFim(char entrada[]);
void getResp(bool resp);
bool isVogal(char n);
bool isConsoante(char n);
bool isNumero(char n);
bool checkApenasVogais(char entrada[]);
bool checkInt(char entrada[]);
bool checkApenasConsoantes(char entrada[]);
bool checkFloat(char entrada[]);

//funcao principal
int main(void){
  char entrada[MAXTAM];
  do{
    scanf(" %[^\n]*c", entrada);
    if(!checkFim(entrada)){
       getResp(checkApenasVogais(entrada));
       getResp(checkApenasConsoantes(entrada));
       getResp(checkInt(entrada));
       getResp(checkFloat(entrada));
       printf("\n");
    } 
  }while(!checkFim(entrada));
  return 0; 
}

//funcao que determina o fim do programa
bool checkFim(char entrada[]){
  return (entrada[0] == 'F' && entrada[1] == 'I' && entrada[2] == 'M');
}

//funcao responsavel por printar a resposta
void getResp(bool resp){
  if(resp == true)
    printf("SIM ");
  else
    printf("NAO ");
}
//funcao auxiliar que checa vogais
bool isVogal(char n){
  bool resp = false;
  char vogais[] = {'a','e','o','u','A','E','O','U'};

  for(int i = 0; i < 10; i++){
    if(n == vogais[i]){
      resp = true;
      i = 10;
    }
  }
  return resp;
}

//funcao auxiliar que checa consoantes
bool isConsoante(char n){
  bool resp = false;
  if(isVogal(n) == false && (n > 65 && n <= 90 || n > 97 && n <= 122))
    resp = true;
  return resp;
}

//metodo auxiliar que verifica algarismos
bool isNumero(char n){
  bool resp = false;
  if(n >= 48 && n <= 57)
    resp = true;
  return resp;
}
/* 4 funcoes da questao */

//checagem do primeiro topico
bool checkApenasVogais(char entrada[]){
  bool resp = true;
  for(int i = 0; i < strlen(entrada); i++){
    if(isVogal(entrada[i]) == false && entrada[i] != ' '){
      resp = false;
      i = strlen(entrada);
    }
  }
  return resp;
}
bool checkApenasConsoantes(char entrada[]){
  bool resp = true;
  for(int i = 0; i < strlen(entrada); i++){
    if(!isConsoante(entrada[i]) && entrada[i] != ' '){
      resp = false;
      i = strlen(entrada);
    } 
  }
  return resp;
}
//checagem do terceiro topico
bool checkInt(char entrada[]){
  bool resp = true;
  for(int i = 0; i < strlen(entrada); i++){
    if(!isNumero(entrada[i]) && entrada[0] != '-'){
      resp = false;
      i = strlen(entrada); 
    }
  }
  return resp;
}

//checagem do ultimo topico
bool checkFloat(char entrada[]){
  bool resp = true;
  int contadorVirgula = 0, contadorPonto = 0;
  int tam = strlen(entrada);

  if(!checkInt(entrada)){
    for(int i = 0; i < tam; i++){
      if(!(isNumero(entrada[i])) && (entrada[i] != '.') && (entrada[i] != ',')){
        resp = false;
        i = tam;
      }
      else if(entrada[i] == ','){
        if(contadorVirgula > 0){
          resp = false;
          i = tam;
        }
        else{
          contadorVirgula++;
        }
      }
      else if(entrada[i] == '.'){
        if(contadorPonto > 0){
          resp = false;
          i = tam;
        }
        else{
          contadorPonto++;
        }
      }
    }
  }
  return resp;
}

File: 2ndSemester/AlgorithimsAndDataStructuresII/Practice/PraticalAssignment01/Q06-IsJava/Is.java
class Is{
  //metodos auxiliares que evitam repeticao de codigo
  public static String getResposta(boolean entrada){
    String resp;
    if(entrada)
      resp = "SIM";
    else 
      resp = "NAO";
    return resp;
  }
  
  public static boolean isVogal(char entrada){
      boolean resp = false;
      char []vogais = {'a','e','i','o','u'};

      for(int i = 0; i < vogais.length; i++){
        if(entrada == vogais[i]){
          resp = true;
          i = vogais.length;
        } 
      }
      return resp;
  }

  public static boolean isConsoante(char entrada){
    boolean resp = false;
      if(isVogal(entrada) == false && (entrada > 65 && entrada <= 90 || entrada > 97 && entrada <= 122)){
        resp = true;
      }
      return resp;
  }

  public static boolean isNumero(char entrada){
    boolean resp = false;
    if ((entrada >= 48 && entrada <= 57)){
      resp = true;
    }
    return resp;
  }

  //quatro metodos da questao
  public static boolean checkApenasVogal(String entrada){
    boolean resp = true;
    for(int i = 0; i < entrada.length(); i++){
      if(isVogal(entrada.charAt(i)) == false && entrada.charAt(i) !=' '){
        resp = false;
        i = entrada.length();
      }
    }  
    return resp;
  }

  public static boolean checkApenasConsoante(String entrada){
    boolean resp = true;
    for(int i = 0; i < entrada.length(); i++){
      if(isConsoante(entrada.charAt(i)) == false && entrada.charAt(i) != ' '){
        resp = false;
        i = entrada.length();
      }
    }
    return resp;
  }
  
  public static boolean checkInteiro(String entrada){
    boolean resp = true;
    for(int i = 0; i < entrada.length(); i++){                 //checagem de algarismo ou sinal de -
      if(!isNumero(entrada.charAt(i)) && entrada.charAt(0) != '-'){
        resp = false;
        i = entrada.length();
      }
    }
    return resp;
  }
 
  public static boolean checkReal(String entrada){
    boolean resp = true;
    int contadorVirgula = 0, contadorPonto = 0;

    if(checkInteiro(entrada) == false){
      for(int i = 0; i < entrada.length();i++){               //checagem de algarismo ou sinais - ou .
        if(!(isNumero(entrada.charAt(i))) && (entrada.charAt(i) != '.') && (entrada.charAt(i) != ',')){
          resp = false;
          i = entrada.length();
        }
        else if(entrada.charAt(i) ==','){
           if(contadorVirgula > 0){ 
             resp = false;
             i = entrada.length();
           }
           else {
            contadorVirgula++;
           }
        }
        else if(entrada.charAt(i) == '.'){
          if(contadorPonto > 0){
            resp = false;
            i = entrada.length();
          }
          else {
            contadorPonto++;
          }
        }
      }
    }
    return resp;   
  }

  public static boolean checkFim(String entrada){
    return (entrada.length() == 3 && entrada.charAt(0) == 'F' && entrada.charAt(1) == 'I' && entrada.charAt(2) == 'M');
  }
  
  //metodo principal
  public static void main(String[] args){
    String entrada;
    do{
      entrada = MyIO.readLine();
      if(checkFim(entrada) == false){
          boolean apenasVogais = checkApenasVogal(entrada),
                  apenasConsoantes = checkApenasConsoante(entrada),
                  valorInteiro = checkInteiro(entrada),
                  valorReal = checkReal(entrada);

          MyIO.println(getResposta(apenasVogais) + " " + getResposta(apenasConsoantes) + " " + getResposta(valorInteiro) + " " + getResposta(valorReal));
      }
    }while(checkFim(entrada) == false);
  } 
}
